RecoLocalTracker/SiPixelRecHits/plugins/SiPixelRecHitCUDA.cc

```diff
@@ -0,0 +1,109 @@
+#include "RecoLocalTracker/Records/interface/TkPixelCPERecord.h"
+#include "RecoLocalTracker/SiPixelRecHits/interface/PixelCPEBase.h"
+#include "RecoLocalTracker/SiPixelRecHits/interface/PixelCPEFast.h"
+
+#include "PixelRecHits.h"  // TODO : spit product from kernel
+
+class SiPixelRecHitCUDA : public edm::global::EDProducer<> {
+public:
+  explicit SiPixelRecHitCUDA(const edm::ParameterSet& iConfig);
+  ~SiPixelRecHitCUDA() override = default;
+
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+private:
+  void produce(edm::StreamID streamID, edm::Event& iEvent, const edm::EventSetup& iSetup) const override;
+
+  // The mess with inputs will be cleaned up when migrating to the new framework
+  edm::EDGetTokenT<cms::cuda::Product<BeamSpotCUDA>> tBeamSpot;
+  edm::EDGetTokenT<cms::cuda::Product<SiPixelClustersCUDA>> token_;
+  edm::EDGetTokenT<cms::cuda::Product<SiPixelDigisCUDA>> tokenDigi_;
+
+  edm::EDPutTokenT<cms::cuda::Product<TrackingRecHit2DCUDA>> tokenHit_;
+
+  std::string cpeName_;
+
+  pixelgpudetails::PixelRecHitGPUKernel gpuAlgo_;
+};
+
+SiPixelRecHitCUDA::SiPixelRecHitCUDA(const edm::ParameterSet& iConfig)
+    : tBeamSpot(consumes<cms::cuda::Product<BeamSpotCUDA>>(iConfig.getParameter<edm::InputTag>("beamSpot"))),
+      token_(consumes<cms::cuda::Product<SiPixelClustersCUDA>>(iConfig.getParameter<edm::InputTag>("src"))),
+      tokenDigi_(consumes<cms::cuda::Product<SiPixelDigisCUDA>>(iConfig.getParameter<edm::InputTag>("src"))),
+      tokenHit_(produces<cms::cuda::Product<TrackingRecHit2DCUDA>>()),
+      cpeName_(iConfig.getParameter<std::string>("CPE")) {}
+
+void SiPixelRecHitCUDA::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+
+  desc.add<edm::InputTag>("beamSpot", edm::InputTag("offlineBeamSpotCUDA"));
+  desc.add<edm::InputTag>("src", edm::InputTag("siPixelClustersCUDAPreSplitting"));
+  desc.add<std::string>("CPE", "PixelCPEFast");
+  descriptions.add("siPixelRecHitCUDA", desc);
+}
+
+void SiPixelRecHitCUDA::produce(edm::StreamID streamID, edm::Event& iEvent, const edm::EventSetup& es) const {
+  // const TrackerGeometry *geom_ = nullptr;
+  const PixelClusterParameterEstimator* cpe_ = nullptr;
+
+  /*
+  edm::ESHandle<TrackerGeometry> geom;
+  es.get<TrackerDigiGeometryRecord>().get( geom );
+  geom_ = geom.product();
+  */
+
+  edm::ESHandle<PixelClusterParameterEstimator> hCPE;
+  es.get<TkPixelCPERecord>().get(cpeName_, hCPE);
+  cpe_ = dynamic_cast<const PixelCPEBase*>(hCPE.product());
+
+  PixelCPEFast const* fcpe = dynamic_cast<const PixelCPEFast*>(cpe_);
+  if (!fcpe) {
+    throw cms::Exception("Configuration") << "too bad, not a fast cpe gpu processing not possible....";
+  }
+
+  edm::Handle<cms::cuda::Product<SiPixelClustersCUDA>> hclusters;
+  iEvent.getByToken(token_, hclusters);
+
+  cms::cuda::ScopedContextProduce ctx{*hclusters};
+  auto const& clusters = ctx.get(*hclusters);
```

Comment:
It would compile but the type of `clusters` would become `cms::cuda::Product<SiPixelClustersCUDA>` instead of current `SiPixelClustersCUDA`.

In the future I hope to enable something along
```cpp
// cms::cuda::ScopedContextProduce (or equivalent) provided by the framework
auto const& clusters = ctx.get(iEvent.get(token_));
// or possibly also
auto const& clusters = ctx.get(iEvent, token_);
```